src/network/message.rs

```diff
@@ -37,21 +37,64 @@ use consensus::encode::MAX_VEC_SIZE;
 #[derive(PartialEq, Eq, Clone, Debug)]
 pub struct CommandString(Cow<'static, str>);
 
+impl CommandString {
+    /// Convert from various string types into a [CommandString].
+    ///
+    /// Supported types are:
+    /// - `&'static str`
+    /// - `String`
+    /// - `&str`
+    ///
+    /// Returns an empty error if and only if the string is
+    /// larger than 12 characters in length.
+    pub fn from<S: Into<Cow<'static, str>>>(s: S) -> Result<CommandString, ()> {
+        let cow = s.into();
+        if cow.as_ref().len() > 12 {
+            Err(())
+        } else {
+            Ok(CommandString(cow))
+        }
+    }
+}
+
 impl fmt::Display for CommandString {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.write_str(self.0.as_ref())
     }
 }
 
+impl str::FromStr for CommandString {
+    type Err = ();
+    fn from_str(s: &str) -> Result<CommandString, ()> {
+        // Don't use CommandString::from to avoid allocating when
+        // it's not needed.
+        if s.len() > 12 {
+            Err(())
+        } else {
+            Ok(CommandString(s.to_owned().into()))
+        }
+    }
+}
+
+#[deprecated(since = "0.26.0", note = "use CommandString::from")]
 impl From<&'static str> for CommandString {
     fn from(f: &'static str) -> Self {
-        CommandString(f.into())
+        if f.len() > 12 {
+            CommandString(f[0..12].into())
```

Comment:
I don't like this, I'd rather remove the `From` impl or panic here then do unexpected things

Reply:
oh I see you deprecated it, hmm idk what's the right thing to do
also, deprecating a trait impl doesn't do anything UX wise:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c67bc53c2372c636208adbfab79fa02e
https://github.com/rust-lang/rust/issues/39935

Reply:
Hmm, that seems to be a cargo bug. I would prefer not to panic, so perhaps just removing the method is the best option.

Reply:
K I pushed a new commit that cleans up the CommandString API. I thin it's such a infrequently used type that all the fancy construction methods are not needed. So I removed all of them.